src/ReTestItems.jl

```diff
@@ -194,47 +211,71 @@ function runtests end
 
 default_shouldrun(ti::TestItem) = true
 
-runtests() = runtests(default_shouldrun, dirname(Base.active_project()))
-runtests(dir::String, dirs::String...) = runtests(default_shouldrun, dir, dirs...)
+runtests(; kw...) = runtests(default_shouldrun, dirname(Base.active_project()); kw...)
+runtests(dir::String, dirs::String...; kw...) = runtests(default_shouldrun, dir, dirs...; kw...)
 
-runtests(pkg::Module) = runtests(default_shouldrun, pkg)
-function runtests(shouldrun, pkg::Module)
+runtests(pkg::Module; kw...) = runtests(default_shouldrun, pkg; kw...)
+function runtests(shouldrun, pkg::Module; kw...)
     dir = pkgdir(pkg)
     isnothing(dir) && error("Could not find directory for `$pkg`")
-    return runtests(shouldrun, _src_and_test(dir)...)
+    return runtests(shouldrun, _src_and_test(dir)...; kw...)
 end
 
-chan(ch::RemoteChannel) = channel_from_id(remoteref_id(ch))
-
-# FilteredChannel applies a filtering function `f` to items
-# when you try to `put!` and only puts if `f` returns true.
-struct FilteredChannel{F, T}
-    f::F
-    ch::T
+function runtests(shouldrun, dirs::String...; verbose=false)
+    if verbose > 0
+        LoggingExtras.withlevel(LoggingExtras.Debug; verbosity=verbose) do
+            _runtests(shouldrun, dirs...)
+        end
+    else
+        return _runtests(shouldrun, dirs...)
+    end
 end
 
-Base.put!(ch::FilteredChannel, x) = ch.f(x) && put!(ch.ch, x)
-Base.take!(ch::FilteredChannel) = take!(ch.ch)
-Base.close(ch::FilteredChannel) = close(ch.ch)
-Base.close(ch::FilteredChannel, e::Exception) = close(ch.ch, e)
-Base.isopen(ch::FilteredChannel) = isopen(ch.ch)
-
-function runtests(shouldrun, dirs::String...)
-    setupctx = SetupContext()
+function _runtests(shouldrun, dirs::String...)
     # Don't recursively call `runtests` e.g. if we `include` a file which calls it.
     get(task_local_storage(), :__RE_TEST_RUNNING__, false) && return nothing
-    ch = RemoteChannel(() -> Channel{TestItem}(Inf))
-    # run test file including @spawn so we can immediately
-    # start executing tests as they're put into our test channel
-    fch = FilteredChannel(shouldrun, chan(ch))
-    @debug "Including tests in $dirs."
-    errormonitor(@spawn include_testfiles!($(setupctx.setups), $fch, dirs...))
-    @debug "Scheduling tests on $(nprocs()) processes."
-    if nprocs() == 1
-        schedule_testitems!(setupctx, chan(ch))
-    else
-        futures = [@spawnat(p, schedule_testitems!(setupctx, ch)) for p in workers()]
-        foreach(wait, futures)
+    # identify project directory for which we're running tests
+    projectfile = ""
+    for dir in dirs
+        path = dir
```

Comment:
probably need an `abspath` somewhere, else we could hit `dirname("")` forever... maybe here?
```suggestion
        path = abspath(dir)
```